src/main/scala/dpla/ingestion3/harvesters/oai/OaiResponseProcessor.scala

```diff
@@ -60,20 +80,102 @@ object OaiResponseProcessor {
     * @return Seq[String]
     *         The set ids.
     */
-  def getSetIdsFromRecord(xml: NodeSeq): Seq[String] = {
+  def getSetIdsFromRecord(xml: Node): Seq[String] = {
     val sets = xml \\ "setSpec"
     sets.map(s => s.text)
   }
 
   /**
-    * Get the error property if it exists
+    * Parses sets from an OaiSource, and filter if necessary according to a
+    * whitelist or blacklist.
     *
-    * @return Option[String]
-    *         The error code if it exists otherwise empty string
+    * @param source OaiSource
+    *               The single-page response to a single OAI query
+    *
+    * @param setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]])
+    *                   A partial function indicating how sets should be filtered,
+    *                   ie. according to a whitelist or blacklist.
+    */
+  def getSets(source: OaiSource, setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]]): OaiResponse = {
+
+    // Parse all sets from XML.
+    def getSetsFromXml(xml: Node): Seq[OaiSet] = {
+      val xmlSets = xml \\ "OAI-PMH" \\ "set"
+
+      xmlSets.flatMap(set => set.headOption match {
+        case Some(node) => {
+          val id = getSetIdentifier(node)
+          Some(OaiSet(id, node.toString, source))
+        }
+        case _ => None
+      })
+    }
+
+    val text = source.text.getOrElse("")
+
+    getXml(text) match {
+      case Failure(e) => {
+        OaiError(e.toString, source)
+      }
+      case Success(xml) => {
+        val sets = getSetsFromXml(xml)
+        // Apply partial function to filter out any sets that do not match some criteria.
+        val filtered = setFilter(sets)
+        SetsPage(filtered)
+      }
+    }
+  }
+
+  /**
+    * Accepts a set from an OAI feed an returns the OAI identifier
+    *
+    * @param set Node
+    *            The original set from the OAI feed
+    * @return The local identifier
+    */
+  def getSetIdentifier(set: Node): String = {
```

Comment:
I'm not going to have any curly braces left! ;)  If I were a dragon, I would be jealously guarding a gleaming pile of golden curly braces.